Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm

```diff
@@ -287,103 +266,156 @@
     if (m_pendingSamples.isEmpty() || m_isDecodingSample)
         return;
 
-    auto pair = m_pendingSamples.takeFirst();
-    decodeSample(pair.first.get(), pair.second);
+    m_isDecodingSample = true;
+    auto tuple = m_pendingSamples.takeFirst();
+    decodeSample(std::get<RetainPtr<CMSampleBufferRef>>(tuple).get(), std::get<bool>(tuple))->whenSettled(m_decompressionQueue, [weakThis = ThreadSafeWeakPtr { *this }, this, flushId = std::get<uint32_t>(tuple)](auto&& result) {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis || isInvalidated())
+            return;
+        assertIsCurrent(m_decompressionQueue.get());
+        --m_framesBeingDecoded;
+        m_isDecodingSample = false;
+        m_lastDecodingError = noErr;
+        m_lastDecodedSample = nullptr;
+
+        if (!result) {
+            ensureOnMainThread([protectedThis = Ref { *this }, this, status = result.error(), flushId] {
+                assertIsMainThread();
+                if (!m_errorListener || flushId != m_flushId)
+                    return;
+                m_errorListener(status);
+            });
+        } else {
+            if (*result) {
+                if (!m_deliverDecodedFrames) {
+                    m_enqueingQueue->dispatch([protectedThis = Ref { *this }, imageSampleBuffer = WTFMove(*result), flushId] {
+                        if (flushId == protectedThis->m_flushId)
+                            protectedThis->enqueueDecodedSample(imageSampleBuffer.get());
+                    });
+                } else {
+                    RunLoop::main().dispatch([protectedThis = Ref { *this }, this, imageSampleBuffer = WTFMove(*result), flushId]() mutable {
+                        assertIsMainThread();
+                        if (flushId == m_flushId && m_newDecodedFrameCallback) {
```

Comment:
Why can m_flushId be checked from either main thread or m_enqueingQueue?

Reply:
m_flushId is an atomic
it all depends on where the end result occurs.
with decodedFrameCallback ; it's called on the main thread, so this is where flushId is checked against.

For the enqueuing of frames, this occurs on the enqueuing thread, and so this is where flushed is checked against.